### DISCUSSION TITLE
Pkey methods PRF and HKDF only use the provider API

### DISCUSSION THREAD

**Discussion by ShuaiYuan21** [openssl/openssl#21627] [Q&A] [answered]
In OpenSSL 3.0, the function tls1_PRF() will just fetch the default provider to do KDF, but there is no judgment for the scene using the engine API. Is it possible to add support for engine API? (HKDF also has the same issue)

For example, first judge whether it is a legacy case, if yes, call the engine API, otherwise call the provider API.

OpenSSL 3.0: 
https://github.com/openssl/openssl/blob/5c56cef501c5344a0e1c1061b356e74c24ca1cef/ssl/t1_enc.c#L48-L72
OpenSSL 1.1.1: 
https://github.com/openssl/openssl/blob/fe824ce0c5d51e7e7cf36c31db6c49c1c0c04a25/ssl/t1_enc.c#L41-L50

**Comment by paulidale** [comment]
I'm not entire sure what you are asking about here.  The second code segment should still work, it's deprecated and won't be removed until at least 4.x but it should still work.  HKDF isn't supported via the PKEY interface and never will be.

The project has been pretty clear since 3.0 was released: there will not be any further enhancements to the engine APIs.  They are deprecated and providers are the replacement.

**Reply by ShuaiYuan21** [reply]
Sorry for not being very clear. 

Taking tls1.2 as an example, in OpenSSL 3.0, the _tls1_PRF()_ function will be called during the handshake phase, and this function will only get the provider through _EVP_KDF_fetch()_ to complete the KDF operation, instead of calling the _EVP_PKEY_CTX_XXX_ series functions like openssl1.1.1.

Since OpenSSL 3.0 also supports the legacy engine API, this function should make one judgment to see if it is a legacy case. If so, then we call the _EVP_PKEY_CTX_XXX_ series of functions to use the engine API, otherwise it will be completed with the help of provider through _EVP_KDF_fetch_

**Comment by t8m** [comment, marked as answer]
The problem is how to find out when the legacy API should be used and when it shouldn't. In theory we could revert to always using the legacy API however it could be seen as functionality change as that will go through different code paths in providers (perhaps even third party ones). And there is no clear way how to infer that an engine providing the legacy EVP_PKEY_CTX style API for HKDF or TLS1 PRF is enabled. I am afraid this ship has sailed and the calls will not be changed apart from some hacks like adding a new SSL_OP to enable the legacy way.
